### DISCUSSION TITLE
handling inconsistent JSON in PowerBI response

### DISCUSSION THREAD

**Discussion by smcclure17** [covid-projections/can-scrapers#256] [Scraper help]
_This is probably better to discuss on a zoom/google meet as it's a lot, but I wanted to post the questions here first._

I'm working on county-level demographics for South Dakota (see PR here #254), which uses a PowerBI dashboard. I've been able to grab the data I want, but in some cases the JSON structure is inconsistent. 

### Context / the Issues:
Each county returns a JSON object that contains a breakdown of each type of dose (1st dose, 2nd dose) by the manufacturer and the demographics. For example in the below snippet of a response, `DM3` contains all the types of doses by manufacturer, and `DM5` contains a list of all the demographics and their values. 

![image](https://user-images.githubusercontent.com/55333380/116286882-bd9e5700-a75d-11eb-8f4b-85c19ddd35ff.png)

This format works great, I'm able to grab everything I need. But, in some cases, the JSON is returned is a bit different (as shown below). I think (but am far from confident) that the `"R"` might stand for "Repeated" or "is the same as the last record". So, if Gregory has 0 `moderna 2 dose` for `asian`, `black`,  and `hispanic`, it will return 0 for `asian` (the first in the list) and then "repeated" for the next two. 

_this snippet is taken from Gregory county_
![image](https://user-images.githubusercontent.com/55333380/116295468-20e0b700-a767-11eb-8cbb-e3a71a99d49c.png)

The consequence of this, with our current method of parsing the JSON, is that all entries with these `"R"` values are stored as `NaNs`, (which, without knowing what the values actually are is probably a good thing). These `NaNs` are carried over into the calculation of the `total_vaccine_initiated` and `total_vaccine_completed` data, meaning we lose those variables for counties with this `NaN` data. 

_Below, Minnehaha has no "R" entries, so all the data is collected, but Gregory does, so the variables are lost for `black`, `hispanic`, and `unknown` demographics._
![image](https://user-images.githubusercontent.com/55333380/116291413-8bdbbf00-a762-11eb-9f05-5f710c3fa05e.png)

To show the prevalence of this, the table below shows how many variables are returned from the `race` scraper for each county. Each county should return 12 variables (vaccine initiated and completed for all 6 races), but most return fewer or none at all.  

![image](https://user-images.githubusercontent.com/55333380/116290229-44a0fe80-a761-11eb-8af6-13849886202d.png)
 
### Questions / Problems:
- First, I'm uncertain what the `"R"` entries/values actually mean. Them being "repeated" values is my best guess, but I'm not sure how to verify/validate it, as they do not surface this county-level data. 
- Second, if this is the case, I'm having trouble thinking of a way to "look back" at previous values/entries with our current system--although I haven't put too much time into this. 
- Third, does this matter? Is this worth pursuing, or should we just report the counties that don't have this issue? It does not seem persistent, in that counties that had "R" values yesterday do not necessarily appear to have them today, but it does appear to affect smaller counties more often (I'm assuming because of repeated "0" values.)

**Comment by tlyon3** [comment]
If `"R"` does in fact mean "same as last record", then I would guess we would forward-fill (not sure if that's the right term) on DB insertion. The PowerBI API, I don't think, has a way to request historical records. So if we see a county missing an entry for a previously used demographic, I would fill it with the most recent known data.

**Reply by smcclure17** [reply]
I think that you're right that filling in the most recent known data might be the best idea. 

Just for clarity, I think the `"R"` means repeated data from the variable before it, not its _own_ past data. 
`SDVaccineCounty` scraper has the same issue/structure (but to a _much_ lesser extent), so for example, looking at Mellette county:

Today, there is the same number of `pfizer completed` and `pfizer 1 dose` values. So, the JSON returns the data (4) for the `pfizer 1 dose`, and then a 'repeated' flag/value for the `pfizer completed`, indicating that the `completed` value is the same as the `1 dose` value. 

![image](https://user-images.githubusercontent.com/55333380/116449225-f276e080-a827-11eb-84ab-1c082bb6771f.png)
![image](https://user-images.githubusercontent.com/55333380/116448849-8d22ef80-a827-11eb-970c-4134facd3063.png)

**Reply by tlyon3** [reply]
Gotcha. If you follow your logic, can you get the _total_ dose numbers to add up?

**Reply by smcclure17** [reply]
total dose across which variables? The response does give the total dose numbers for each manufacturer/dose combination, which matches the data on the dashboard. When all the demographic breakdowns are present (no `"R"s`) the summation of each demographic does match this total value.

**Reply by tlyon3** [reply]
Hmm...maybe I'm not completely following, but if you can get the total dose numbers for each demographic to line up with the dashboard, I'd say the scraper is good

**Reply by smcclure17** [reply]
all good, I am not explaining well at all

**Comment by sglyon** [comment]
Could we test the meaning of "R" by 

1. doing the forward fill (based on order of results in the json arrays returned by API)
2. Computing all categories for a particular dimension (e.g. sex in [male, female, unknown]) and manufacturer + dose number
3. Summing across the categories for the dimension and seeing if it matches the total doses for manufacturer + dose number as reported near top of dashboard?

**Reply by smcclure17** [reply]
yep, will start on that now
